Automatically fix name clashes on instantiation in editor

FroznMine opened the pull request:
This is a very basic fix for #26390.

When instancing dependent scenes in the editor the clashing names are now automatically changed and an error is logged. The renaming is done during PackedScene/SceneState::instance but only when the scene is opened in the editor. I currently did this only for the editior, as I think that automatic renaming while in-game would cause even more confusion to the user. (Compared to the current state of having multiple nodes with the same path/name.)

### Caveats

It is not very obvious to the user that renaming took place as this is only logged as an error in the console and the scene is not displayed as modified [(*) marking in scene tab is missing]. 
The renaming needs to be implemented somewhere else to solve these issues.

### Different Approaches
While trying to fix the issue my first intuition was to check for name clashes while loading the scene through ResourceLoader. During resource loading dependency errors can occur. These dependency errors are currently all 'missing dependency'-error. These can be resolved using the DependencyEditor, if they occur during loading. As the name duplication also is some type of dependency error - a 'name clash with dependency'-error - my idea was to generalize the notion of dependency errors and extend the DependencyEditor with a way to resolve name clashes.

However, this approach has two complications:
1. ResourceLoaders only create a PackedScene with SceneState in which it is very difficult to detect name clashes. As SceneStates only have a representation of the serialized scene, no nodes are instanced. Hence the node hierarchy is not immediately traversable, since all nodes are just `int` identifiers with another int-id referencing their parents. Finding name clashes would require to reimplement a node tree traversal on the level of SceneStates/PackedScenes.

2. This approach would not show effects when the dependent scene is currently opened. When switching to the dependent scene (as done e.g. in #28133) the editor uses `EditorData::check_and_update_scene(int p_idx)` to refresh the scene and show the newly added node. This is done by packing and reinstancing the current scene:
https://github.com/godotengine/godot/blob/080b5df6256a368fcd767aec96dd2200194542c8/editor/editor_data.cpp#L642-L645
Hence the resource loader is not involved there and the check would not be performed when implemented in the resource loaders. Thus the solution must be implemented in the instantiation too (code duplication) or else we need to change the way the scene is updated. The last one might be possible by writing the packed scene to a temporary file and load from this instead. But this in turn could have other side effects which need to be considered and handled.

So this approach is probably too much work.

Here is another idea: Implementing more detailed error handling for PackedScene instancing.
I noticed that the ResourceLoader class has static callbacks to inform about errors during loading:
https://github.com/godotengine/godot/blob/080b5df6256a368fcd767aec96dd2200194542c8/core/io/resource_loader.h#L162-L176

As far as I can tell `notify_load_error/err_notify` is used nowhere and `notify_dependency_error/dep_err_notify` is only set by the EditorNode. This allows the engine to show the DependencyEditor to fix dependencies when in-editor and letting loading fail immediately with a (logged) error when in-game. Implementing a similar approach for PackedScene/SceneState on instancing of scenes would allow the editor to provide better guidance for the user in case of problems (like name clashes introduced by dependencies) while still simply failing (with a logged error) in-game.

While requiring more in-depth work on SceneState, EditorNode and the creation of a NameClashResolver (similar to DependencyEditor) I think that this approach would give the best user experience. One might also be able to use it to make other editor errors more prominent to the user.

Another way to solve the problem would be moving the automatic renaming into the EditorNode where it is performed as part of loading and updating of scenes. Here it can be implemented as a series of undo/redo operations and hence correctly show the scene as modified. However this still does not solve the problem of properly informing the user about the renaming.

I would like to hear your opinions on the possible solutions.

Fixes #26390
_Bugsquad edit:_ Fixes #14680 too

FroznMine committed:
Automatically fix name clashes on instantiation in editor

Simple fix for #26390.

Adding a node to a scene which is inherited or instanced elsewhere
can result in name clashes in dependent scenes. When instancing such
dependent scenes in the editor the clashing names are automatically changed
and an error is logged.

asheraryam commented:
I prefer that this happens during the updating/save of the scene, since that makes it less likely that it will go unnoticed.

I don't like the idea of this happening while the project is running, because how do you decide which node to rename? I could be calling get_node("AnimPlayer") and it can break randomly without me understanding why.

If we don't want to make it automatic in the editor, the simplest fix for now is to add a warning next to the nodes with the same name.

I prefer we add a warning as a start and plan the auto-renaming more thoroughly.

FroznMine commented:
> How do you decide which node to rename? 

In this PR the last node added which creates a collision is automatically renamed during instantiation. As the children of instanced/inherited scene nodes are added first they will never get renamed. Only the custom children added to the instanced node will get renamed. This is consistent with the [reasoning of santouits](https://github.com/godotengine/godot/issues/26390#issuecomment-489371322) in #26390 that name clashes are the problem of the inheriting/instancing scene instead of the inherited/instanced scene.

> I could be calling get_node("AnimPlayer") and it can break randomly without me understanding why.

I agree with this. The 'solution' in this PR can cause even more confusion than the current situation, as the node renaming happens without much explicit information for, and interaction with the user. Thats why I proposed the implementation of more detailed error handling for PackedScene instancing as a better solution. When such clashes occur the editor should provide the user with an interface where all duplicate node names together with all (local) references of said names are displayed and the user can perform the renaming. (Think git merge conflict resolution.) Finding all the references is probably the biggest problem, however I think there is already something similar implemented for the moving of nodes in the SceneTree. I also just saw that there exists a RenameDialog which could probably be useful for that.

> The simplest fix for now is to add a warning next to the nodes with the same name.

I wanted to do this instead of printing the error to the console first. However, currently the only way to show a warning in the scene tree is to have the node produce a non empty string in its get_configuration_warning() method. Hijacking this to display name clashes is probably not a good idea, since many nodes use it to display their specific warnings. A name clash might even stay unnoticed due to this. Maybe one could implement this through a new 'validity warning' symbol?

asheraryam commented:
I think we need both the auto-renaming and the error handling.

My reasoning is that the auto-renaming for the editor should solve most issues naturally, since users are used to having names changed automatically by adding a number.

However, it could still be possible to edit the scene file manually and force names to be duplicates, so additional error handling would be very welcome.

I don't have issues with the PR, it can be merged after more testing and the error handling can be added in a later PR.

FroznMine committed:
Print error on node name clashes in-game

Prints an error as in the editor, if a node has two children with the same name.
However, no nodes are renamed if this occurs while running a game to prevent
unexpected behaviour for the user.

FroznMine commented:
> However, it could still be possible to edit the scene file manually and force names to be duplicates, so additional error handling would be very welcome.

Oh, I see what you mean. I misunderstood you there.

I added another commit in which an error will always be logged, independent of being in the editor or not, but the node is only renamed if the name clash happens in the editor. If undesired I can remove this commit from the PR.

asheraryam commented:
LGTM 👍

Anutrix commented:
Since https://github.com/godotengine/godot/issues/26390 is closed as completed, is this still applicable/needed?
